src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/GooglePubSubReceiver.php

```diff
@@ -0,0 +1,146 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Bridge\GooglePubSub\Transport;
+
+use Google\Cloud\PubSub\Message;
+use Google\Cloud\PubSub\Subscription;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Exception\LogicException;
+use Symfony\Component\Messenger\Exception\MessageDecodingFailedException;
+use Symfony\Component\Messenger\Transport\Receiver\ReceiverInterface;
+use Symfony\Component\Messenger\Transport\Serialization\PhpSerializer;
+use Symfony\Component\Messenger\Transport\Serialization\SerializerInterface;
+
+/**
+ * @author Richard Henkenjohann <richardhenkenjohann@googlemail.com>
+ */
+class GooglePubSubReceiver implements ReceiverInterface
+{
+    private const MESSAGE_ATTRIBUTE_NAME = 'X-Symfony-Messenger';
+
+    private $subscription;
+    private $serializer;
+
+    /** @var array[] */
+    private $buffer = [];
+
+    public function __construct(Subscription $subscription, SerializerInterface $serializer = null)
+    {
+        $this->subscription = $subscription;
+        $this->serializer   = $serializer ?? new PhpSerializer();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function get(): iterable
+    {
+        $gpsEnvelope = $this->getMessage();
+        if (null === $gpsEnvelope) {
+            return;
+        }
+
+        try {
+            $envelope = $this->serializer->decode([
+                'body' => $gpsEnvelope['body'],
+                'headers' => $gpsEnvelope['headers'],
+            ]);
+        } catch (MessageDecodingFailedException $exception) {
+            $this->subscription->delete($gpsEnvelope['id']);
+
+            throw $exception;
+        }
+
+        yield $envelope->with(new GooglePubSubReceivedStamp($gpsEnvelope['id'], $gpsEnvelope['ackId']));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function ack(Envelope $envelope): void
+    {
+        $stamp = $this->findReceivedStamp($envelope);
+
+        $this->subscription->acknowledge(new Message([], ['ackId' => $stamp->getAckId()]));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function reject(Envelope $envelope): void
+    {
+        $this->ack($envelope);
```

Comment:
This has to be wrong. You can NACK message by modifying ack deadline to 0 as it's documented here https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/modifyAckDeadline.

Reply:
Ok, that's a great point! Thanks for bringing this up.
So basically `modifyAckDeadline(0)` will allow other subscribers to handle the message while `ack()` will permanently remove the message (which is not what we want in a Pub/Sub setting)